lib/Sema/PerformanceHints.cpp

```diff
@@ -52,6 +67,88 @@ void checkImplicitCopyReturnType(const ClosureExpr *Closure,
   }
 }
 
+class CheckExistentialAny : public TypeVisitor<CheckExistentialAny, bool> {
+public:
+  static bool inType(Type type) {
+    return CheckExistentialAny().visit(type->getCanonicalType());
+  }
+
+  static void inFunctionReturnType(FuncDecl *FD, DiagnosticEngine &Diags) {
+    Type T = FD->getResultInterfaceType();
+
+    if (inType(T))
+      Diags.diagnose(FD, diag::perf_hint_func_returns_existential_any, FD);
+  }
+
+  static void inClosureReturnType(ClosureExpr *CE, DiagnosticEngine &Diags) {
+    Type T = CE->getResultType();
+
+    if (inType(T))
+      Diags.diagnose(CE->getLoc(),
+                     diag::perf_hint_closure_returns_existential_any);
+  }
+
+  static void inVariableType(const VarDecl *VD, DiagnosticEngine &Diags) {
+    Type T = VD->getInterfaceType();
+
+    if (inType(T))
+      Diags.diagnose(VD, diag::perf_hint_var_uses_existential_any, VD);
+  }
+
+  static void inPatternType(const AnyPattern *AP, DiagnosticEngine &Diags) {
+    Type T = AP->getType();
+
+    if (inType(T))
+      Diags.diagnose(AP->getLoc(),
+                     diag::perf_hint_any_pattern_uses_existential_any);
+  }
+
+  static void inTypeAlias(const TypeAliasDecl *TAD, DiagnosticEngine &Diags) {
+    Type T = TAD->getUnderlyingType();
+
+    if (inType(T))
+      Diags.diagnose(TAD->getLoc(),
+                     diag::perf_hint_typealias_uses_existential_any, TAD);
+  }
+
+  bool visitExistentialType(ExistentialType *ET) {
+    return true;
+  }
+
+  bool visitTupleType(TupleType *TT) {
```

Comment:
Using TypeVisitor for this is incorrect because you're going to miss a lot of cases. For example, a nominal type might have a parent, which you're not visiting at all:

```
struct Outer<T> {
  struct Inner {}
}

Outer<any P>.Inner
```

Please use Type.findIf() or Type.visit() instead.